### ISSUE TITLE
Swarm mode in 17.09.0-ce not replace completed/failed tasks as designed

### ISSUE THREAD

**Issue by ozlevka-ericom** [moby/moby#35080]
Hi there,

We have case when we not use hard coded replicas in service, but we have service that keep number of replicas by our own logic instead. With upgrade to 17.09.0-ce from 17.06.2-ce we found that our logic stop working instead design:
By documentation of task behavior, failed tasks should:

[If the task fails the orchestrator removes the task and its container and then creates a new task to replace it according to the desired state specified by the service](https://docs.docker.com/engine/swarm/how-swarm-mode-works/services/#tasks-and-scheduling).

In docker:

```
Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:11 2017
 OS/Arch:      linux/amd64
 Experimental: true
 ```

The behavior as designed:
Step to reproduce:
Create docker-compose.yml:

```yml
version: "3.3"
services:
  server:
    image: nginx
    ports:
      - 80:80
    deploy:
      mode: replicated
      replicas: 0
      restart_policy:
        condition: none
```

run `docker swarm init && docker stack deploy -c docker-compose.yml test && docker && docker service ls`

output is:
```
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
yb6v98o31knb        test_server         replicated          0/0                 nginx:latest        *:80->80/tcp
```

then `docker service scale test_server=5`

`docker service ps test_server` return:

```
ID                  NAME                IMAGE               NODE                      DESIRED STATE       CURRENT STATE           ERROR               PORTS
 m28sw9fjtdrd        test_server.1       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 8 seconds ago
 wdgprv7a4qns        test_server.2       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 8 seconds ago
 u5qbnm2178oa        test_server.3       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 8 seconds ago
 ieydt40t9dlx        test_server.4       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 7 seconds ago
 0bstwmuh1400        test_server.5       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 6 seconds ago
 ```

 Let's kill two containers by `docker rm $(docker kill [container id]`

 The tasks now in status `failed`:

 ```
 ID                  NAME                IMAGE               NODE                      DESIRED STATE       CURRENT STATE           ERROR                         PORTS
  m28sw9fjtdrd        test_server.1       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 2 minutes ago
  wdgprv7a4qns        test_server.2       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 2 minutes ago
  u5qbnm2178oa        test_server.3       nginx:latest        lev-hp-elitebook-850-g3   Shutdown            Failed 5 seconds ago    "task: non-zero exit (137)"
  ieydt40t9dlx        test_server.4       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 2 minutes ago
  0bstwmuh1400        test_server.5       nginx:latest        lev-hp-elitebook-850-g3   Shutdown            Failed 16 seconds ago   "task: non-zero exit (137)"
  ```

Restart tasks status by running: `docker service scale test_server=6`

Then behavior as documented the failed tasks replaced by new one:

```
ID                  NAME                IMAGE               NODE                      DESIRED STATE       CURRENT STATE                    ERROR                         PORTS
 m28sw9fjtdrd        test_server.1       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 4 minutes ago
 wdgprv7a4qns        test_server.2       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 4 minutes ago
 kogxnct8wrgj        test_server.3       nginx:latest        lev-hp-elitebook-850-g3   Running             Running less than a second ago
 u5qbnm2178oa         \_ test_server.3   nginx:latest        lev-hp-elitebook-850-g3   Shutdown            Failed 2 minutes ago             "task: non-zero exit (137)"
 ieydt40t9dlx        test_server.4       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 4 minutes ago
 6sljck2m855c        test_server.5       nginx:latest        lev-hp-elitebook-850-g3   Running             Running 1 second ago
 0bstwmuh1400         \_ test_server.5   nginx:latest        lev-hp-elitebook-850-g3   Shutdown            Failed 2 minutes ago             "task: non-zero exit (137)"
 cydglz4o314b        test_server.6       nginx:latest        lev-hp-elitebook-850-g3   Running             Starting 2 seconds ago
 ```

In docker:

```
Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
 ```

 After return on previouse steps tasks still in `failed` state:

 ```
 ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE                ERROR                         PORTS
  fbe6l06dwazv        test_server.1       nginx:latest        ubuntu-server       Running             Running about a minute ago
  v9bxxa6ur3kp        test_server.2       nginx:latest        ubuntu-server       Running             Running about a minute ago
  ox0852jr0kie        test_server.3       nginx:latest        ubuntu-server       Shutdown            Failed 17 seconds ago        "task: non-zero exit (137)"
  dhxttorbjb5e        test_server.4       nginx:latest        ubuntu-server       Running             Running about a minute ago
  cy30hiqppags        test_server.5       nginx:latest        ubuntu-server       Shutdown            Failed 28 seconds ago        "task: non-zero exit (137)"
  rm1sxbojsdkf        test_server.6       nginx:latest        ubuntu-server       Running             Running 3 seconds ago
 ```

 Instead designed.

 If scale service to 4 and scale back to 6 then failed tasks is deleted.

 Thank you for your help..

**Comment by ozlevka-ericom** [comment]
Please change version/17.06 to version/17.09 label

**Comment by thaJeztah** [comment]
I don't see this being a bug; you create a service with `restart-policy: none`, which tells swarm that when a task fails, no attempt should be taken to start a new task.

In your situation, you are manually killing the task (which gets detected as a failure), and Swarm updates the desired state to "Shutdown" (i.e.; you told it to not keep the task running when it fails, so the desired state is updated to "Shutdown" to reflect that).

In general, I would really not recommend manually interacting (stopping/restarting/killing) with containers (tasks)  that are part of a service; doing so can lead to undesired side-effects, for example restarting a task leads to swarm creating a _new_ container to replace the old one, and stop monitoring the old one (thus: two instances of the container running, one of which is unmanaged).

I'm closing this because I don't think there's a bug at hand, but feel free to continue the conversation if I misunderstood.

**Comment by ozlevka-ericom** [comment]
All functionality should be as dockumented.
If you think that it's not bug so please fix documentation with exception on this situation.
Some thing like: "restart_policy should be great from zero"
But you can't broke functionality with out any description or warning...
17.06 work and 17.09 not... Some warning or breaking changes document...

**Comment by ErezPasternak** [comment]
We have an interesting use case, this why we have this configuration 

Beside "normal" services we need a special one, the best way to describe it is a disposable one time usage container

This container can be in 3 stats , Ready ( so it can be used by a single remote user ) , in Use , Removed ( after usage is done )
In order to have good performance , we are making sure that we will always at least X containers in Ready stat, and they are used, we issue a service scale command to get more containers ( Container Pool Management ) 
We need to have between 10 and 10K containers on this type  , this led us to use Swarm due to the multi machine support.

For this logic, we have set the restart-policy: none , and the container itself once done is exiting, and until the last Docker version, this exited container Task was remove as well, so a scale command after this case worked,

With the new version , while swarm show for example 25/30 , a scale command to 30 does not start 5 new containers, it seems that Swarm see 25 containers and 30 tasks
If you have a different way to handle such use case, please share

**Comment by thaJeztah** [comment]
Reopening to discuss the behaviour, but I suspect the change in behaviour was due to a bug fix, and the old behaviour was relying on a bug

ping @aaronlehmann @nishanttotla @anshulpundir PTAL

**Comment by ErezPasternak** [comment]
I think that this new behavior is a bug since
In our case 25/30 scale command to 30 will not scale , while scale to 25 and then to 30 will work

I think scale to 30 should work , am I wrong ?

**Comment by thaJeztah** [comment]
> In our case 25/30 scale command to 30 will not scale

If I understand correctly; the service has already `--replicas=30`, but there's 5 tasks that have existed (so 25/30)? If the current definition of the service is already `--replicas=30`, then no update takes place in the service's spec, so no new tasks will be deployed.

> while scale to 25 and then to 30 will work

In that case you're forcing an update of the service, but changing a property (`--replicas`), then changing it back

**Comment by thaJeztah** [comment]
Wondering though, why you're using `restart-policy: none`? From your description:

> the container itself once done is exiting

A task's lifecycle in Swarm can only go in one direction (created -> running -> exited -> removed), _never_ goes back, irregardless the restart policy. So if you have a container that automatically exits  once it's job is done, you can probably achieve what you want by just setting the default restart-policy.

In that case;

container (task) starts -> container is used -> container exits -> swarm starts a new task to replace it

**Comment by ozlevka-ericom** [comment]
Correct.
Update --replicas 25
And second update --replicas 30.
Working correct. Instead previouse version when any change return completed task to run.

On Oct 4, 2017 4:56 PM, Sebastiaan van Stijn <notifications@github.com> wrote:

In our case 25/30 scale command to 30 will not scale

If I understand correctly; the service has already --replicas=30, but there's 5 tasks that have existed (so 25/30)? If the current definition of the service is already --replicas=30, then no update takes place in the service's spec, so no new tasks will be deployed.

while scale to 25 and then to 30 will work

In that case you're forcing an update of the service, but changing a property (--replicas), then changing it back

**Comment by ozlevka-ericom** [comment]
Because our logic can scale down...
In this case exists chanse to kill container that serve users...

On Oct 4, 2017 5:02 PM, Sebastiaan van Stijn <notifications@github.com> wrote:

Wondering though, why you're using restart-policy: none? From your description:

the container itself once done is exiting

A task's lifecycle in Swarm can only go in one direction (created -> running -> exited -> removed), never goes back, irregardless the restart policy. So if you have a container that automatically exits once it's job is done, you can probably achieve what you want by just setting the default restart-policy.

In that case;

container (task) starts -> container is used -> container exits -> swarm starts a new task to replace it

**Comment by thaJeztah** [comment]
So, summarising the behaviour you're expecting:

- A task for a service with `--restart-condition=none` should not be restarted (replaced by a new task) when it exits
- When scaling _down_ the service, exited tasks should be removed _first_, and running tasks should only be stopped/removed if the number of desired tasks (`--replicas`) is lower than the number of running tasks (this, to prevent tasks that are in use by a customer from being killed)
- When running `docker service scale`; all stopped tasks should be "restarted" (new tasks started to replace them), _even_ if the number of replicas _before_ scaling is the same as the number of replicas _after_ scaling (basically, running `docker service scale` should reset the desired state, and restart monitoring those tasks; even though `--restart-condition=none` was set)
- scaling up or down should not restart (replace) tasks that are currently running

Does that summarise your expectations correctly?

**Comment by ozlevka-ericom** [comment]
Not exactly:
* A task for a service with --restart-condition=none should not be restarted (replaced by a new task) when it exits => this is correct
* When scaling down the service, exited tasks should be removed first. => *already exists* and work correctly in 17.09 to.
* Running tasks should only be stopped/removed if the number of desired tasks (--replicas) is lower than the number of running tasks (this, to prevent tasks that are in use by a customer from being killed). It's not exactly correct. We never scale down if all tasks in running status. 
_We scale down if gap between running and completed tasks exists only. Then this is less important._
* When running docker service scale; all stopped tasks should be **"killed/removed"** (and new tasks started to replace them), even if the number of replicas before scaling is the same (_I not sure about word "same" because is not exactly same_)  as the number of replicas after scaling (basically, running docker service scale should reset the desired state, and restart monitoring those tasks; even though --restart-condition=none was set). I would sugest **restart_policy=manual** for this kind of functional.
* scaling up or down should not restart (replace) tasks that are currently running. _Seems this functionality already exists and correct_...

**Comment by thaJeztah** [comment]
> We scale down if gap between running and completed tasks exists only. Then this is less important.

Understood; if (e.g.) status is `25/30`, and you scale down from 30 to 25, no running tasks would be killed. If, however, you would scale down to 20, then 5 "already stopped" tasks are removed, and 5 still running tasks would be killed.

>  all stopped tasks should be "killed/removed" 

That would depend on the `--task-history-limit` that is configured for the swarm; setting `--task-history-limit` to 0 would mean "when a task exits, remove it", instead of keeping failed tasks

> I not sure about word "same" because is not exactly same

Well, this is basically the problem in this issue; the `docker service scale` (and `docker service update` commands) are designed to be _idempotent_; for example, running `docker service scale=5` multiple times should not make any modification to the running service, and the _state_ of the service (because there's no change in the service specs).

The use case here though is to _do_ make changes to the service's state, even if there's no change (i.e., restart tasks, even though the service spec says to not restart them if they failed).

While I understand the use case, swarm's implementation of services is not really designed for that (also see https://github.com/moby/moby/issues/23880), and I'm not sure that behaviour should be expected. Perhaps a `docker service start`, or `docker task xyz` commands would be more suitable for that, but I haven't given that one a lot of thought.

**Comment by aaronlehmann** [comment]
> Well, this is basically the problem in this issue; the `docker service scale` (and `docker service update` commands) are designed to be _idempotent_; for example, running `docker service scale=5` multiple times should not make any modification to the running service, and the _state_ of the service (because there's no change in the service specs).

Yes, I agree with this statement.

The difference in behavior is probably due to https://github.com/docker/swarmkit/pull/2332. This fixed what I consider a longstanding bug that any update to the service was causing all tasks to be restarted, even if the restart policy was set to "none".

I think one way to accomplish what you want would be to temporarily change the restart condition to "any", scale the service, and then change the restart condition back to "none".

**Comment by ozlevka-ericom** [comment]
The problem is not that is idemponet and not change status. We can change to x+1 or x-1.
The problem is "zombie" tasks that take slots of service and right now avoid creating new "healthy" tasks

**Comment by thaJeztah** [comment]
I think that would still be possible, but requires you to be more explicit to indicate you want to get rid of old tasks;

Create a "one-off" service with 5 replicas:

```bash
$ docker service create --name one-off --restart-condition=none --replicas=5 nginx:alpine

$ docker service ps one-off
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE            ERROR               PORTS
37ojcqs7l92j        one-off.1           nginx:alpine        moby                Running             Running 11 seconds ago
rq7wqms9wraq        one-off.2           nginx:alpine        moby                Running             Running 11 seconds ago
jb4ohv0ud974        one-off.3           nginx:alpine        moby                Running             Running 11 seconds ago
jw5dp6dkoqm0        one-off.4           nginx:alpine        moby                Running             Running 11 seconds ago
tq9wu9u4dsap        one-off.5           nginx:alpine        moby                Running             Running 11 seconds ago
```

Kill two tasks to simulate they have completed their job

```bash
$ docker kill one-off.1.37ojcqs7l92jx2en2xa6fh1b1 one-off.5.tq9wu9u4dsapvqu22dlk40m2l

$ docker service ps one-off

ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE                ERROR                         PORTS
37ojcqs7l92j        one-off.1           nginx:alpine        moby                Shutdown            Failed 4 seconds ago         "task: non-zero exit (137)"
rq7wqms9wraq        one-off.2           nginx:alpine        moby                Running             Running about a minute ago
jb4ohv0ud974        one-off.3           nginx:alpine        moby                Running             Running about a minute ago
jw5dp6dkoqm0        one-off.4           nginx:alpine        moby                Running             Running about a minute ago
tq9wu9u4dsap        one-off.5           nginx:alpine        moby                Shutdown            Failed 4 seconds ago         "task: non-zero exit (137)"

$ docker service ls

ID                  NAME                MODE                REPLICAS            IMAGE                   PORTS
nj9jjph53gxj        one-off             replicated          3/5                 nginx:alpine
```

From the above, 2 out of 3 tasks are no longer running, so can be disposed, so:

Remove the old tasks:

```bash
$ docker service scale --detach=true one-off=3
```

Then scale back to spin up the amount of "ready" tasks you want:

```bash
$ docker service scale --detach=true one-off=5
```

After which, 5 tasks are running again:

```bash
$ docker service ps one-off

ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE            ERROR               PORTS
rs1pkwnetxf2        one-off.1           nginx:alpine        moby                Running             Running 33 seconds ago
rq7wqms9wraq        one-off.2           nginx:alpine        moby                Running             Running 3 minutes ago
jb4ohv0ud974        one-off.3           nginx:alpine        moby                Running             Running 3 minutes ago
jw5dp6dkoqm0        one-off.4           nginx:alpine        moby                Running             Running 3 minutes ago
i6fl4iymx2l9        one-off.5           nginx:alpine        moby                Running             Running 33 seconds ago
```

**Comment by ozlevka-ericom** [comment]
Thank you...
We sure consider this workarround.
But for us it's still workarround...


On Oct 6, 2017 1:03 PM, Sebastiaan van Stijn <notifications@github.com> wrote:

I think that would still be possible, but requires you to be more explicit to indicate you want to get rid of old tasks;

Create a "one-off" service with 5 replicas:

$ docker service create --name one-off --restart-condition=none --replicas=5 nginx:alpine

$ docker service ps one-off
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE            ERROR               PORTS
37ojcqs7l92j        one-off.1           nginx:alpine        moby                Running             Running 11 seconds ago
rq7wqms9wraq        one-off.2           nginx:alpine        moby                Running             Running 11 seconds ago
jb4ohv0ud974        one-off.3           nginx:alpine        moby                Running             Running 11 seconds ago
jw5dp6dkoqm0        one-off.4           nginx:alpine        moby                Running             Running 11 seconds ago
tq9wu9u4dsap        one-off.5           nginx:alpine        moby                Running             Running 11 seconds ago

Kill two tasks to simulate they have completed their job

$ docker kill one-off.1.37ojcqs7l92jx2en2xa6fh1b1 one-off.5.tq9wu9u4dsapvqu22dlk40m2l

$ docker service ps one-off

ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE                ERROR                         PORTS
37ojcqs7l92j        one-off.1           nginx:alpine        moby                Shutdown            Failed 4 seconds ago         "task: non-zero exit (137)"
rq7wqms9wraq        one-off.2           nginx:alpine        moby                Running             Running about a minute ago
jb4ohv0ud974        one-off.3           nginx:alpine        moby                Running             Running about a minute ago
jw5dp6dkoqm0        one-off.4           nginx:alpine        moby                Running             Running about a minute ago
tq9wu9u4dsap        one-off.5           nginx:alpine        moby                Shutdown            Failed 4 seconds ago         "task: non-zero exit (137)"

$ docker service ls

ID                  NAME                MODE                REPLICAS            IMAGE                   PORTS
nj9jjph53gxj        one-off             replicated          3/5                 nginx:alpine

From the above, 2 out of 3 tasks are no longer running, so can be disposed, so:

Remove the old tasks:

$ docker service scale --detach=true one-off=3

Then scale back to spin up the amount of "ready" tasks you want:

$ docker service scale --detach=true one-off=5

After which, 5 tasks are running again:

$ docker service ps one-off

ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE            ERROR               PORTS
rs1pkwnetxf2        one-off.1           nginx:alpine        moby                Running             Running 33 seconds ago
rq7wqms9wraq        one-off.2           nginx:alpine        moby                Running             Running 3 minutes ago
jb4ohv0ud974        one-off.3           nginx:alpine        moby                Running             Running 3 minutes ago
jw5dp6dkoqm0        one-off.4           nginx:alpine        moby                Running             Running 3 minutes ago
i6fl4iymx2l9        one-off.5           nginx:alpine        moby                Running             Running 33 seconds ago

**Comment by cpuguy83** [comment]
I think in any case you are doing a work-around as swarm does not really support this use case.
Although it's not really clear why you can't use swarm's reconciliation for replicas and are instead relying on your own.

**Comment by thaJeztah** [comment]
@cpuguy83 IIUC, for this use case the problem was scaling _down_; swarm's reconciliation would automatically spin up new instances (this may be desirable), but when scaling down, there's no easy way to tell which instances are currently in use by a customer, and which ones are "free"  (thus can be removed upon scaling down)

**Comment by cpuguy83** [comment]
Ah. Sounds like swarmkit could use some method assisted scaling.
